src/main/java/com/contrastsecurity/exceptions/HttpResponseException.java

```diff
@@ -0,0 +1,150 @@
+package com.contrastsecurity.exceptions;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Thrown when the Contrast API returns a response that indicates an error has occurred. Captures
+ * the HTTP status line and body.
+ *
+ * <p>Typically, HTTP status codes in the 4XX and 5XX ranges indicate an error; however, in some
+ * legacy cases, the Contrast API may return a successful status code (e.g. 200) with an error body.
+ * In this case, the Contrast Java SDK still throws this exception.
+ */
+public class HttpResponseException extends ContrastException {
+
+  /**
+   * Static factory method that captures the status line and body of the given {@code
+   * HttpURLConnection} to create a new exception.
+   *
+   * @param connection connection from which to derive status line and body information
+   * @param message error message provided by the caller
+   * @return new {@code HttpResponseException}
+   * @throws IOException when fails to read the status line or body of the response
+   */
+  public static HttpResponseException fromConnection(
+      final HttpURLConnection connection, final String message) throws IOException {
+    final int code = connection.getResponseCode();
+    final ExceptionFactory factory = factoryFromResponseCode(code);
+    final String body = readBody(connection, code);
+    final String status = connection.getResponseMessage();
+    return factory.create(message, code, status, body);
+  }
+
+  /**
+   * Maps HTTP response codes to the constructor for the appropriate concrete type.
+   *
+   * @param code response code
+   * @return factory for creating a new concrete type of {@code HttpResponseException}
+   */
+  private static ExceptionFactory factoryFromResponseCode(final int code) {
+    switch (code) {
+      case 401:
+      case 403:
+        return UnauthorizedException::new;
+      case 404:
+        return ResourceNotFoundException::new;
+      default:
+        return HttpResponseException::new;
+    }
+  }
+
+  /**
+   * Reads the response body in its entirety. Error responses should always be relatively small, so
+   * they should fit into memory.
+   *
+   * @return body as a string, or {@code null}
+   */
+  private static String readBody(final HttpURLConnection connection, final int code)
```

Comment:
This kinda sucks. You think TeamServer will ever use proper response codes?

Reply:
I think you're referring to the legacy TeamServer code that returns 200 and an error object in the response body. This code is actually not related to that. In that case, the code making the request would have already had to read the body to determine if an exception needs to be thrown.

This method is used when the code making the request has determined that it doesn't like the status code returned by the API, and it wants to throw a meaningful exception. This code reads debug information from the body of the response to help you understand why the request failed. For example TeamServer might return:

```
HTTP/2 400
date: Fri, 20 Aug 2021 15:53:31 GMT
content-type: application/json
content-length: 46

{"error":{"message":"Missing application ID"}}
```

If the `HttpResponseException` _only_ had the status code and some message like "failed to read vulnerabilities", the developer reading the exception would have no way to know the error message returned in the response body. This code makes sure the developer gets that information so they don't have to do anything painful like put Wireshark between the SDK and TeamServer to debug.

Reply:
I'm testing with the Maven plugin now, and I'm finding that I'd like even more debug information in the exception message. I'm going to add the request path and method.